src/main/groovy/com/blackbaud/templates/GitRepo.groovy

```diff
@@ -33,7 +37,36 @@ class GitRepo {
     }
 
     void pushProject(String username, String password) {
-       git.push().setCredentialsProvider(new UsernamePasswordCredentialsProvider(username, password)).setPushAll().call()
+        Iterable<PushResult> results = git.push().setCredentialsProvider(new UsernamePasswordCredentialsProvider(username, password)).setPushAll().call()
+        for (PushResult result : results) {
+            parsePushResult(result)
+        }
+    }
+
+    private static void parsePushResult(PushResult result) {
+        for (RemoteRefUpdate remoteRefUpdate : result.remoteUpdates) {
+            if(isRemoteRefUpdateStatusFailed(remoteRefUpdate.status)) {
+               throw new GitPushFailedException(remoteRefUpdate.message)
+            }
+        }
+    }
+
+    private static boolean isRemoteRefUpdateStatusFailed(RemoteRefUpdate.Status status) {
+        switch (status) {
+            case OK:
+            case UP_TO_DATE:
+            case NON_EXISTING:
+                return false;
+            case NOT_ATTEMPTED:
+            case AWAITING_REPORT:
+            case REJECTED_NODELETE:
+            case REJECTED_NONFASTFORWARD:
+            case REJECTED_REMOTE_CHANGED:
+            case REJECTED_OTHER_REASON:
+                return true;
+            default:
+                return true;
```

Comment:
if a previously unseen status is returned do you want this method to return true like it is now, or blow up about it not knowing about that previously un-accounted for status?

Reply:
seems like you can get rid of the above cases and just return false on the three and true on everything else.  in the case of a previously unknown state, the state will either indicate update status failed or did not fail.  if you explode on unknown, then you have to always change the code.  if you return true on unknown, then you only have to change the code if the previously unknown state is a non-status failed state (which seems less likely than a new status failed state)

Reply:
true will treat it as a failure and make an exception raise, which i think is preferable to it silently failing